DRAFT: Prometheus: Improve handling of large amounts of metric names

s-h-a-d-o-w opened the pull request:
**What this PR does / why we need it**:
Metric names will now be fetched via web worker, not blocking the UI thread. Also, the connection will be dropped once we have received 100K metric names.

**Which issue(s) this PR fixes**:
Fixes #21893

**Special notes for your reviewer**:
- You can use [this mock server](https://github.com/s-h-a-d-o-w/prometheus-mock) if you want to see the effects of this in action.
- I also tried out `worker-plugin` but ended up not using it because it threw warnings when it tried to process a library that we use that already uses web workers ([a known issue](https://github.com/GoogleChromeLabs/worker-plugin/issues/12)): `[Webpack] WARNING in new Worker() will only be bundled if passed a String.`
It also makes [eslint process already compiled code](https://github.com/webpack-contrib/eslint-loader/issues/261#issuecomment-587236828), obviously resulting in lots of errors.
- Raised the metric name threshold to 100K because we can easily handle that in terms of metric names.

**Minor additional improvements**
- Used to be that metadata was mapped to metrics on every key stroke, even though they are only fetched in `start()`. Now, they are mapped right there, in `start()`. This uses more RAM but obviously improves performance. Theoretically, metrics and metadata could be removed if it wasn't for the fact that they're used in dependencies - where that on-the-fly mapping happens as well. So... it seems like things could be improved further by only having the "premapped" metrics data. But - significant refactoring task.
- Using a worker for `GroupMetricsByPrefix` addresses a minor hitch once all data actually arrived. (With a large data set, UI would freeze for a second or so.)

**Minor downsides**
- Requests take longer to execute because of the overhead from streaming. It's negligible with just hundreds or thousands of metric names though. And beyond that, obviously not blocking the main thread becomes more important.
- `eslint-loader` currently doesn't appear to work with workers. But since we lint on CI, I think that this is also negligible.
- No solution for TS import yet (note the `@ts-ignore` for the import) - official docs [only explain loader syntax](https://github.com/webpack-contrib/worker-loader#integrating-with-typescript) and when I tried the same thing without it, TS simply wouldn't find the type for the module. Hence the @ts-ignore.

s-h-a-d-o-w committed:
Prometheus: Improve handling of large amounts of metric names

Metric names will now be fetched via web worker, not blocking
the UI thread. Also, the connection will be dropped once we
have received 100K metric names.

Fixes #21893

aocenas commented on public/app/plugins/datasource/prometheus/language_provider.ts:
```diff
@@ -111,11 +115,54 @@ export default class PromQlLanguageProvider extends LanguageProvider {
 
+  requestViaWorker(url: string, hasObjectResponse?: false): Promise<string[]>;
+  requestViaWorker(url: string, hasObjectResponse?: true): Promise<PromMetricsMetadata>;
+  requestViaWorker(url: string, hasObjectResponse?: boolean): Promise<string[] | PromMetricsMetadata>;
```
Seems like this override is not needed if we have false/true already.

s-h-a-d-o-w replied:
Unfortunately, it is necessary for proper autocomplete support. If this line is removed, you will only see 2 out of the 3 possible signatures, as TS doesn't add the implementation signature automatically.

s-h-a-d-o-w replied:
I should maybe elaborate a bit more on this - the crucial thing is that it would work as it is used right now but if `requestViaWorker()` was to be called with a boolean variable as the second argument, you'd get:
`Argument of type 'boolean' is not assignable to parameter of type 'true'.ts(2769)`

aocenas commented on public/app/plugins/datasource/prometheus/language_provider.ts:
```diff
@@ -111,11 +115,54 @@ export default class PromQlLanguageProvider extends LanguageProvider {
+      };
+
+      streamWorker.onerror = error => {
+        resolve(nodes);
```
Shouldn't we reject the promise with error in that case? Maybe having an Error with nodes property, so that component can then decide what to do with it, like ideal I would assume show an error toast while updating the metrics so app can continue working but at least user knows not everything is fine.

s-h-a-d-o-w replied:
I was just working off of the previous implementation - which was to just return the default if an error happens and the error (such as 405 with prometheus metadata) only gets logged to the console.

aocenas replied:
Got it but still, we should do something with the error, right now we are just silently ignoring it, which is the most frustrating thing when debugging some weird behavior.

s-h-a-d-o-w replied:
Actually, the button does say "(No metadata found)".

aocenas commented on public/app/plugins/datasource/prometheus/workers/StreamJSONResponse.worker.ts:
```diff
@@ -0,0 +1,95 @@
+  postMessage(message: StreamJSONResponsePayload['data'], options?: PostMessageOptions): void;
+}
+
+export function streamJSONResponse(data: StreamJSONResponsePayload['data'], callback: (arg: any) => void) {
```
Nitpick: Maybe we can change to `callback: (arg: JSON_STREAM_DONE | any)` would not be much of help with type checking but would add some context for reader.

s-h-a-d-o-w replied:
Good catch! I've improved on it further by defining types for `ArrayChunk` and `ObjectChunk`.

aocenas commented on public/app/plugins/datasource/prometheus/language_provider.ts:
```diff
@@ -111,11 +115,54 @@ export default class PromQlLanguageProvider extends LanguageProvider {
-    this.lookupsDisabled = this.metrics.length > this.lookupMetricsThreshold;
-    this.metricsMetadata = await this.request('/api/v1/metadata', {});
-    this.processHistogramMetrics(this.metrics);
+    const [metrics, metricsMetadata] = await Promise.all([
```
Does it make sense to make this cancel-able? We wrap it her ehttps://github.com/grafana/grafana/blob/a4b7c60c2ba5f86ea2ee10b19e24a851e95447c9/public/app/plugins/datasource/prometheus/components/PromQueryField.tsx#L123 with makePromiseCancelable but that just ignores the result. In the worker we check if we are already fetching and so cannot start fetching again. If for some reason the fetching still takes long (slow connection or something) user needs to reload the page cause just navigating somewhere else probably won't work is that right? Maybe we can add something so that this can be cancelled on component unmount which would call worker.terminate()

s-h-a-d-o-w replied:
Since `start()` is async, I don't see how it would block anything aside from things that are related to the collection of metrics data.

But... just to be safe, I will try to mock a response that takes forever to resolve and see how this behaves - maybe oboe triggers a timeout anyway. And also how it used to behave. Because maybe it's important for things that depend on it that `start()` simply doesn't finish in such a case.

s-h-a-d-o-w replied:
I just looked at the current behavior - it's the same. And it works quite well - the button keeps saying "Loading metrics..." but since it's async, you can of course keep doing everything else, like executing queries and whatnot.

Cleaning up a pending request after a while if the server doesn't respond - I suppose it's cleaner but since it's an improvement over the current behavior and would require quite a bit of effort on my part, I would suggest to do this improvement separately in the future.

aocenas reviewed (commented):
Nice looks pretty solid, have just a few comments and questions.

s-h-a-d-o-w commented on public/app/plugins/datasource/prometheus/language_provider.ts:
```diff
@@ -111,11 +115,54 @@ export default class PromQlLanguageProvider extends LanguageProvider {
+
+      streamWorker.postMessage({
+        chunkSize: Math.ceil(this.lookupMetricsThreshold / 20),
+        headers: this.datasource.basicAuth ? { Authorization: this.datasource.basicAuth } : { 'X-Grafana-Org-Id': 1 },
```
Yeah that's obviously exactly what I tried to do here. :)
Since how oboe does requests is precisely the point why I chose to use it, one can't just "hand off" the task of executing the request somehow.

aocenas replied:
Well I mean I am not sure for example this part https://github.com/grafana/grafana/blob/40d195e4a71e9ffefb50ad4ed2439df285650790/public/app/core/services/backend_srv.ts#L397 seems like we are considering some headers we are not dealing with in the worker. And even though we may not use all of it with prometheus, may create some errors in the future. So I would rather use some single handleHeaders fucntion that would be used both in the worker and backend_srv to make sure both are handling the headers in the same way.

davkal replied:
We have to guarantee that the request can get through to the datasource, possibly through out datasource proxy. For that it needs to set the Org ID out of the user context, otherwise the request won't get through, or this will fail in other ways in a multi-org installation. I realize that it would be quite neat if the datasource could arrange for its own transport, but there is too much at stake if let each datasource do that independently.
I feel like the workerization should already prevent the sluggishness, and I'd be happy if the user has to wait a bit to see all results. What else do you get from streaming?

s-h-a-d-o-w replied:
@davkal I'm actually just now wrapping up on isolating and exposing the header transformation done by the backend service, should be ready to review in a few minutes. :)

s-h-a-d-o-w committed:
StreamJSONResponse worker: Improve callback type

s-h-a-d-o-w committed:
Expose header transformation done by backend_srv

So that it can be used at other places in the code
where have to provide headers and can't just
reuse the backend_srv request API.

s-h-a-d-o-w commented:
@hugohaggmark @dprokop @torkelo I just before discussed this PR with @davkal and while he appreciates the overall approach and the performance improvement this brings, he has concerns about the architecture regarding the modifications to the backend service API. Which I totally understand and is funnily enough related to the following thing I wanted to do in the first place (why I didn't is another story ;) ).

But David argued that I would have to check with you whether it's alright.
My proposal is that instead of exposing the backend service's header augmentation like I did here, I move the `StreamJSONRespone` worker from the prometheus data source into the backend service. And offer both a `requestStream` and `datasourceRequestStream` that can be used anywhere where one might encounter amounts of data that are so large that fetching them would block the main thread.

What do you think?

torkelo commented:
I have not looked at this PR before, would need some time to dig into and understand the changes to backendSrv and the use of web workers etc. maybe some this could be presented during a frontend weekly as well?

hugohaggmark commented on packages/grafana-runtime/src/services/backendSrv.ts:
```diff
@@ -39,6 +39,13 @@ export interface BackendSrv {
 
   // DataSource requests add hooks into the query inspector
   datasourceRequest(options: BackendSrvRequest): Promise<any>;
+
+  // Can be used to augment the request headers just like the backend service
+  // would anywhere else in the code base. (Obviously, there should be good
+  // reasons not to use the request API that this service provides.)
+  augmentRequestHeaders(headers: BackendSrvRequest['headers']): any;
+
+  augmentDataSourceRequestHeaders(headers: BackendSrvRequest['headers']): any;
```
Not sure this belongs to the public api. Internally we can use the "real" backendSrv like we do here https://github.com/grafana/grafana/blob/02140fff2d15004e7310fa01663610797a3b90b8/public/app/features/search/components/MoveToFolderModal.tsx#L8 and then those methods are accessible.

s-h-a-d-o-w replied:
Fair enough, that's what I've been discussing with @aocenas and @davkal already and is why I'd like to move the worker into the backend service. :)

hugohaggmark commented on public/app/core/services/backend_srv.ts:
```diff
@@ -103,6 +103,34 @@ export class BackendSrv implements BackendService {
     return await this.request({ method: 'PUT', url, data });
   }
 
+  augmentRequestHeaders(headers: BackendSrvRequest['headers']) {
+    let augmentedHeaders: BackendSrvRequest['headers'] = headers;
+    const orgId = this.dependencies.contextSrv.user?.orgId;
+
+    if (orgId) {
+      augmentedHeaders = augmentedHeaders || {};
+      augmentedHeaders['X-Grafana-Org-Id'] = orgId;
+    }
+
+    return augmentedHeaders;
+  }
+
+  augmentDataSourceRequestHeaders(headers: BackendSrvRequest['headers']) {
+    let augmentedHeaders: BackendSrvRequest['headers'] = this.augmentRequestHeaders(headers);
+
+    if (augmentedHeaders?.Authorization) {
+      augmentedHeaders['X-DS-Authorization'] = augmentedHeaders.Authorization;
+      delete augmentedHeaders.Authorization;
+    }
+
+    if (this.noBackendCache) {
+      augmentedHeaders = augmentedHeaders || {};
+      augmentedHeaders['X-Grafana-NoCache'] = 'true';
+    }
+
+    return augmentedHeaders;
+  }
+
```
Great work on the refactoring of this, did you see https://github.com/grafana/grafana/blob/02140fff2d15004e7310fa01663610797a3b90b8/public/app/core/utils/fetch.ts#L4 because we do a lot of header specific work there. Not sure you need it though, just mention it.

s-h-a-d-o-w replied:
Thanks!
IIRC (since all of this is up in the air right now, I won't look at it in detail), I did indeed see this because it is exactly where my "hunt" for headers started and where e.g. `augmentRequestHeaders()` is eventually used if you follow the chain of calls.

hugohaggmark commented:
> @hugohaggmark @dprokop @torkelo I just before discussed this PR with @davkal and while he appreciates the overall approach and the performance improvement this brings, he has concerns about the architecture regarding the modifications to the backend service API. Which I totally understand and is funnily enough related to the following thing I wanted to do in the first place (why I didn't is another story ;) ).
> 
> But David argued that I should check with you whether you're fine with my following proposal.
> Namely that instead of exposing the backend service's header augmentation like I did here, I move the `StreamJSONRespone` worker from the prometheus data source into the backend service. And offer both a `requestStream` and `datasourceRequestStream` that can be used anywhere where one might encounter amounts of data that are so large that fetching them would block the main thread.
> 
> What do you think?

To be honest, I think we should stick with RxJs as our main streaming lib, both backendSrv.request and backendSrv.datasourceRequest could also be called requestStream/datasourceRequestStream but we transform them to Promise because that's how the api worked before we introduced RxJs. But some question to think about concerning, that is:
1. Can we can solve the same issue using RxJs?  
2. If not, can we wrap RxJs around StreamJSONResponse so that the output from requestStream and datasourceRequestStream would be Observable<some type here>?

But these are just the first thoughts that I came up with, I like Torkels idea of presenting this in a FrontEnd weekly to get more details.

dprokop commented:
@s-h-a-d-o-w  I'll have a look at this ltr this week:)

s-h-a-d-o-w commented:
@hugohaggmark At least as far as I've seen, there is no actual streaming happening here, it's just that the methods contain the name `stream` but in the end, it just uses `fetch`, its `.body()` method and `JSON.parse` - none of which streams?
But since you're familiar with rxjs (I'm not), if you could point me towards either a spot in our code or in their docs where actual streaming happens, I'd obviously appreciate it.
Especially since if it turns out that I can do the same thing with rxjs, investing time into a presentation would be unnecessary. :)

hugohaggmark commented:
> @hugohaggmark At least as far as I've seen, there is no actual streaming happening here, it's just that the methods contain the name `stream` but in the end, it just uses `fetch`, its `.text()` method and `JSON.parse` - none of which streams?
> But since you're familiar with rxjs (I'm not), if you could point me towards either a spot in our code or in their docs where actual streaming happens, I'd obviously appreciate it.
> Especially since if it turns out that I can do the same thing with rxjs, investing time into a presentation would be unnecessary. :)

Sorry for being so unclear on the term 'stream'. 

RxJs is a lib that treats any data as a stream of data or you can think of it as an infinite array of data where each index in the aray is a point in time:ish. With backendSrv it wraps `fetch` in an Obeservable with the `fromFetch` function. So from my point of view the result from `fromFetch` is a stream but we don't use it as a stream unfortunately because we call `toPromise` that returns a Promise. We do this because that was the contract from the previous angular api we had.

In the other case, the term `stream` means continous chunks of data transported from the backend to frontend if I understand you correctly. The only place I know we use this in RxJs context is here https://github.com/grafana/grafana/blob/3d17d63891c48226e7d18c064c7738373080687f/public/app/plugins/datasource/loki/datasource.ts#L218 and that is using web sockets to stream chunks.

I can't say if you can use RxJs to solve your specific issue but you can probably wrap the results from the worker into a RxJs stream (Observable) much like RxJs wraps `fetch` with an Observable in `fromFetch` [here](https://github.com/ReactiveX/rxjs/blob/99b5af1af5d169d55d454ff8e27d88105cee4b6f/src/internal/observable/dom/fetch.ts#L104).

That way we don't introduce yet another streaming paradigm. I hope that this clarifies my thoughts :)

s-h-a-d-o-w commented:
@hugohaggmark Thanks for the clarification, I'll look into this!

mckn commented on packages/grafana-ui/src/utils/typeahead.ts:
```diff
@@ -8,8 +8,8 @@ export const flattenGroupItems = (groupedItems: CompletionItemGroup[]): Completi
-    all.push(titleItem, ...items);
-    return all;
+
+    return all.concat(titleItem, items);
```
nit: If we are starting to optimize things for performance and we are talking about really big lists. I would recommend changing from concat to the `apply` pattern.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply#Examples

s-h-a-d-o-w replied:
The reason this was done wasn't primarily performance but the fact that the maximum call stack size is exceeded if `items` is above a certain length. Still - thanks, if this PR moves ahead, I'll evaluate whether `apply` would also work and provides better performance.
